### DISCUSSION TITLE
Behavior of a single sync/async connection while using gather/threads

### DISCUSSION THREAD

**Discussion by CaselIT** [psycopg/psycopg#590] [General]
The current version of psycopg (3.1.9) seems to allow a single async connection to be shared by multiple task by using gather, by serializing access to the db.

```py
import asyncio
import random

import psycopg

# asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())

async def go(conn):
    await asyncio.sleep(random.uniform(0, 0.01))
    print(asyncio.current_task())
    print(await (await conn.execute('SELECT pg_sleep(1)')).fetchone())


async def main():
    conn = await psycopg.AsyncConnection.connect("postgresql://scott:tiger@127.0.0.1:5432/test")
    await asyncio.gather(*[go(conn) for i in range(100)])
```

The same seems to be happening when using a sync connection with thread
```py
from concurrent.futures import ThreadPoolExecutor, as_completed
import random
import threading
import time

import psycopg

def gos(conn):
    time.sleep(random.uniform(0, 0.01))
    print(threading.get_ident())
    print(conn.execute('SELECT pg_sleep(1)').fetchone())


def mains():
    conn = psycopg.Connection.connect("postgresql://scott:tiger@127.0.0.1:5432/test")
    with ThreadPoolExecutor(100) as pool:
        x = [pool.submit(gos, conn) for i in range(100)]
        print([future.result() for future in as_completed(x)])

mains()
```
Both will run the 100 pg_sleep sequentially. This seems surprising to me, I would expect an exception here, indicating that concurrent access was not allowed. If it's a deliberate choice maybe is should be documented. Maybe it's just my misunderstanding of how a shared resource should behave when concurrent access is attempted.

Psyocpg2 seems to have a similar behavior, but it seems to block all threads until all sleeps are done.
Creating a discussion since I'm guess it's a deliberate choice since it seems to be somewhat comparable to what psycopg2 does.

For reference asyncpg running the same example raises an error `asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress`

This came up in sqlalchemy around confusions regarding usage of gather with a single session. https://github.com/sqlalchemy/sqlalchemy/discussions/5980#discussioncomment-6203441. Not really related to this, it's just what prompted be to check what the driver were doing

edit: sorry for the OT, how did you enabled the checkbox before submitting the discussion asking to check for similar ones?

**Comment by dvarrazzo** [comment]
A Postgres connection can execute only one operation at time. It can be executed by more than one Python thread/task, but access will be serialized. This is the expected behaviour.

We will surely not fail a second concurrent access, otherwise the connection object becomes effectively non-thread safe and whoever wants to use it in a multi-thread program will have to use their own lock. Not to mention the massive backwards incompatibility.

Psycopg2 in green mode has an awkward architecture that made impossible to guard the critical section with a lock. The exception you are seeing now is the Postgres connection failing, which is what we avoid in all the other modes.

So, not a bug, a precise choice. The topic "if you want to go weally weally fast you have to use many connections, not many cursors" has been discussed so many time in psycopg life that I just forgot to mention it in the docs. Or I just failed to find it back right now. [In psycopg 2 it is documented](https://www.psycopg.org/docs/usage.html#thread-and-process-safety) and psycopg 3 is not different: there is a lock around the critical section and the serialization comes as a consequence.

https://github.com/psycopg/psycopg/blob/63803637e5b20547807bcd632ce2487d871df44e/psycopg/psycopg/cursor.py#L702-L705

**Reply by CaselIT** [reply]
> A Postgres connection can execute only one operation at time. It can be executed by more than one Python thread/task, but access will be serialized. This is the expected behaviour.

Sure, I'm aware that a connection to a db cannot support multiple operation. I guess I for some reason expected this case to raise an error. Maybe I just encountered it a while back on asyncpg and this morning thought it would be the norm.

> Psycopg2 in green mode has an awkward architecture that made impossible to guard the critical section with a lock. The exception you are seeing now is the Postgres connection failing, which is what we avoid in all the other modes.

using normal python thread did work on psycopg2, but it seemed to block all threads until all pg_sleep were done, so instead of the psycopg 3 behaviour "1s delay, print row, 1s delay, print row, ..." on psycopg2 it's "10s delay, 10 print row". I'm not claiming there is any problem here or anything to be done for psycopg2

>  I just forgot to mention it in the docs. Or I just failed to find it back right now

I can't say I've done an exhaustive search of the docs, but I too did not find it mentioned in the psycopg3 ones

thanks for the reply!
